### PR TITLE
transaction: Resilient Txn Engine

### PR DESCRIPTION
**Pull request by oshankkumar**
The transaction engine executes the given transaction
across the cluster.The engine is designed to make use
of etcd as the means of communication between peers.

Please refer Design Document #1003

Signed-off-by: Oshank Kumar <okumar@redhat.com>

### CODE REVIEW

**Comment by atinmu** [comment]
@oshankkumar Can we please have 'Fixes' or 'Updates' tag in the commit message?

**Comment by atinmu** [comment]
@oshankkumar It might be worth to mention which all parts of #919 are covered here and what's pending. This will give us a good heads up for the reviewers.

**Comment by oshankkumar** [comment]
#### Task List
- [x] Transaction Engine
- [x] Creating and running a transaction.
- [x] Synchronized step execution
- [x] Modify global data structures
- [x] Rollback previously executed steps on all nodes in case of txn failure
- [ ] Cleanup Leader 
- [ ] Handling peer restart during transaction (partially Done)

**Comment by oshankkumar** [comment]
retest this please

File: glusterd2/transaction/context.go [hunk 1]
```diff
@@ -33,29 +33,32 @@ type TxnCtx interface {
 	// Logger returns the Logrus logger associated with the context
 	Logger() log.FieldLogger
 
-	// commit writes all locally cached keys and values into the store using
+	// Commit writes all locally cached keys and values into the store using
```
[the remark this answers is not in the document]
Comment (oshankkumar) [reply glusterd2/transaction/context.go]: Yes Kaushal, I have made few things exported because I wanted to use in new package

**Comment by oshankkumar** [comment]
retest this please

**Comment by oshankkumar** [comment]
retest this please

**Comment by Madhu-1** [comment]
@nigelbabu we have a CI failure, any hint why its failing?
```
13:16:00 [gluster_glusterd2] $ /bin/sh -xe /tmp/jenkins82820089879639684.sh
13:16:00 + SSID_FILE=/home/gluster/workspace/gluster_glusterd2/cico-ssid
13:16:00 ++ cat /home/gluster/workspace/gluster_glusterd2/cico-ssid
13:16:00 cat: /home/gluster/workspace/gluster_glusterd2/cico-ssid: No such file or directory
```

**Comment by oshankkumar** [comment]
retest this please

File: glusterd2/transaction/lock.go [hunk 2]
```diff
@@ -149,15 +149,15 @@ func CreateLockFuncs(key string) (LockUnlockFunc, LockUnlockFunc) {
...
+type Locks map[string]*concurrency.Mutex
+
+func (l Locks) lock(lockID string) error {
 	// Ensure that no prior lock exists for the given lockID in this transaction
-	if _, ok := t.locks[lockID]; ok {
+	if _, ok := l[lockID]; ok {
 		return ErrLockExists
 	}
 
-	logger := t.Ctx.Logger().WithField("lockID", lockID)
```
Comment (kshlm) [inline glusterd2/transaction/lock.go]: Restore the logs here. They can be helpful.

Comment (oshankkumar) [reply glusterd2/transaction/lock.go]: restored the previous logs

**Pull request merged**